api/src/services/setup/router.js

```diff
@@ -0,0 +1,88 @@
+const express = require('express');
+const router = express.Router();
+const path = require('path');
+const localeUtils = require('locale');
+const morgan = require('morgan');
+
+const environment = require('../../environment');
+const startupLog = require('./startup-log');
+
+const { wantsJSON } = require('../../middleware/wants-json');
+const { getLocale } = require('../../middleware/locale');
+const logger = require('../../logger');
+const translations = require('../../translations');
+const config = require('../../config');
+const template = require('../template');
+const branding = require('../../services/branding');
+const STATUS = 503;
+
+router.use(express.static(environment.staticPath));
+router.use(getLocale);
+router.use(
+  morgan('STARTUP RES :remote-addr :remote-user :method :url HTTP/:http-version :status :res[content-length]')
+);
+
+const getEnabledLocales = () => {
+  return translations
+    .getEnabledLocales()
+    .then(docs => docs.map(doc => doc.code))
+    .catch(err => {
+      logger.error('Error getting enabled locales: %o', err);
+      return [];
+    });
+};
+
+const getBestLocaleCode = (acceptedLanguages, locales, defaultLocale) => {
+  const headerLocales = new localeUtils.Locales(acceptedLanguages);
+  const supportedLocales = new localeUtils.Locales(locales, defaultLocale);
+  return headerLocales.best(supportedLocales).language;
+};
+
+const getTemplate = async () => {
+  const templatePath = path.join(environment.templatePath, 'setup', 'setup.html');
+  return template.getTemplate(templatePath);
+};
+
+const renderStartupPage = async (req) => {
+  const enabledLocales = await getEnabledLocales();
+  const locale = getBestLocaleCode(req.locale, enabledLocales, config.get('locale'));
+  const progress = startupLog.getProgress(locale);
+
+  const template = await getTemplate();
+  return template({
+    title: config.translate('api.startup.title', locale, { branding: await branding.get() }),
+    actions: progress.actions,
+    locale: locale,
+  });
+};
+
+router.get('/api/v1/startup-progress', (req, res) => {
+  res.json(startupLog.getProgress(req.locale));
+});
+
+const paths = [
+  '/setup/poll',
+  '/_session',
+  '/api/v2/upgrade',
+  '/api/deploy-info',
+];
+paths.forEach(path => {
+  router.all(path, (req, res) => {
+    res.status(STATUS);
+    res.json({ error: 'Service unavailable' });
+  });
+});
```

Comment:
I think I actually prefer the other approach of being explicit about requesting JSON. It seems more correct to call this out when making a scripting request. Other than "feeling" right there's not much of a difference so either is ok with me.

Reply:
that's what I did the first time around, and then started adding the paths to the router. 
A lot of the reason behind it was that, when I test, the admin app that that actually runs and needs the edit to is the "base" branch that I upgrade from. 

I pretty much agree that the request should explicitly ask for json.

Reply:
Changed!